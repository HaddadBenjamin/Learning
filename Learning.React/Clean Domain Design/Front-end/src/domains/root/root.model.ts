import {Reducer, ReducersMapObject, Store} from "redux";
import { Saga } from "redux-saga";
import {Task} from "@redux-saga/types";

export interface LazyStore extends Store
{
	injectedReducers : ReducersMapObject
	injectedSagas : Map<string, Task>
	
	injectReducer : (key : string, reducer : Reducer) => void
	injectSaga : (key : string, saga : Saga) => void
}