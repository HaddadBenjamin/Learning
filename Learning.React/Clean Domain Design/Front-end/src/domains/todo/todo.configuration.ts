export const routes =
{
    api : '/todos'
}

export const todoReducerKey = 'todos'
export const todoSagaKey = 'todos'