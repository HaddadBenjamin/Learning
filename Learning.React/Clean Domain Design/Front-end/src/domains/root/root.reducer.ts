import { combineReducers } from "redux";
import { todoReducer } from 'domains/todo/todo.reducer'

const rootReducer = combineReducers(
{
    todos : todoReducer
});

export default rootReducer