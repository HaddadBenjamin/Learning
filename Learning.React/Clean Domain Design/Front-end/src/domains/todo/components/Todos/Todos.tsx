import React, { FC, useEffect } from 'react';
import Todo from '../Todo/Todo';
import AddTodo from '../AddTodo/AddTodo';
import { useSelector, useDispatch } from 'react-redux';
import { selectTodos } from '../../todo.selector';
import { getTodosRequestAction } from '../../todo.action';
import styles from './Todos.module.scss'

const Todos : FC = () =>
{
    const todos = useSelector(selectTodos)
    const dispatch = useDispatch()

    useEffect(() => { dispatch(getTodosRequestAction()) }, [])

    return <>
        { todos &&
            <>
                <h1 className={styles.title}>Todo List</h1>
                <div className={styles.container}>
                    { todos && todos.todos.map(t => <Todo {...t} key={t.id}/>) }
                    <AddTodo />
                </div>
            </>
        }
    </>
};

export default Todos;