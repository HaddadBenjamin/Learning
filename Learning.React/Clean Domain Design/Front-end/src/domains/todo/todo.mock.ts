import {TodoState} from "./todo.state";
import {ActionStatus} from "shared/domains/redux/redux.model";

export const emptyTodoStateMock : TodoState =
{
    status : ActionStatus.Loading,
    error : undefined,
    todos : []
}

export const todoStateMock : TodoState =
{
    ...emptyTodoStateMock,
    todos : [
        {
            id : "1",
            title : "Faire le m√©nage",
            completed : false,
        },
        {
            id : "2",
            title : "Faire ma compta",
            completed : true,
        },
        {
            id : "3",
            title : "Appeler Marine",
            completed : false,
        },
        {
            id : "4",
            title : "Acheter des croquettes pour petit chien",
            completed : false,
        },
    ]
}

export const todoMock = todoStateMock.todos[0]
export const {title : todoTitleMock, id : todoIdMock} = todoStateMock.todos[1]
export const errorMessageMock = '404 not found'
export const errorMock  = new Error(errorMessageMock)