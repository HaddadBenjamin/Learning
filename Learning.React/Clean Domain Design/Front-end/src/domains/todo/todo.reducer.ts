import { TodoState, initialTodoState } from './todo.state';
import { TodoAction, TodoActions } from './todo.action';
import { failedActionMetadata, loadedActionMetadata, loadingActionMetadata } from 'shared/domains/redux/redux.util';

export const todoReducer = (state: TodoState = initialTodoState, action: TodoActions)
: TodoState =>
{
    switch (action.type)
    {
        case TodoAction.GET_TODOS_REQUEST :     return loadingActionMetadata(state)
        case TodoAction.GET_TODOS_SUCCESS :     return { ...loadedActionMetadata(state), todos: action.payload }
        case TodoAction.GET_TODOS_FAILED :      return failedActionMetadata(action.error, state)

        case TodoAction.ADD_TODO_REQUEST :      return loadingActionMetadata(state)
        case TodoAction.ADD_TODO_SUCCESS :      return loadedActionMetadata(state)
        case TodoAction.ADD_TODO_FAILED :       return failedActionMetadata(action.error, state)
     
        case TodoAction.EDIT_TODO_REQUEST :     return loadingActionMetadata(state)
        case TodoAction.EDIT_TODO_SUCCESS :     return loadedActionMetadata(state)
        case TodoAction.EDIT_TODO_FAILED :      return failedActionMetadata(action.error, state)

        case TodoAction.TOGGLE_TODO_REQUEST :   return loadingActionMetadata(state)
        case TodoAction.TOGGLE_TODO_SUCCESS :   return loadedActionMetadata(state)
        case TodoAction.TOGGLE_TODO_FAILED :    return failedActionMetadata(action.error, state)

        case TodoAction.REMOVE_TODO_REQUEST :   return loadingActionMetadata(state)
        case TodoAction.REMOVE_TODO_SUCCESS :   return loadedActionMetadata(state)
        case TodoAction.REMOVE_TODO_FAILED :    return failedActionMetadata(action.error, state)

        default : return state
    }
}