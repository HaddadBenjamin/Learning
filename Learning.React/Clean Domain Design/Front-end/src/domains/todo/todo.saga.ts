import { call, put, select, takeLatest } from 'redux-saga/effects'
import {
   addTodoFailedAction,
   AddTodoRequestAction,
   addTodoSuccessAction,
   editTodoFailedAction,
   EditTodoRequestAction,
   editTodoSuccessAction,
   getTodosFailedAction, getTodosRequestAction,
   getTodosSuccessAction,
   removeTodoFailedAction,
   RemoveTodoRequestAction,
   removeTodoSuccessAction,
   TodoAction,
   toggleTodoFailedAction,
   ToggleTodoRequestAction,
   toggleTodoSuccessAction
} from './todo.action';
import {
   addTodo,
   getAllTodos,
   patchTitle,
   removeTodo,
   patchCompleted
} from './todo.api';
import { ITodo } from './todo.model';
import { selectTodos } from './todo.selector';
import { TodoState } from './todo.state';
import {createTodo} from "./todo.repository";

export function* getAllTodosSaga()
{
   try
   {
      const todos : ITodo[] = yield call(getAllTodos);
      yield put(getTodosSuccessAction(todos));
   }
   catch (error)
   {
      yield put(getTodosFailedAction(error.message));
   }
}

export function* addTodoSaga(action : AddTodoRequestAction)
{
   try {
      const newTodo: ITodo = yield call(addTodo, createTodo(action.payload.title));

      yield put(addTodoSuccessAction(newTodo));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(addTodoFailedAction(error.message));
   }
}

export function* editTodoSaga(action : EditTodoRequestAction)
{
   try
   {
      const { id, title } = action.payload
      const editedTodo : ITodo = yield call(patchTitle, id, title);

      yield put(editTodoSuccessAction(editedTodo));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(editTodoFailedAction(error.message));
   }
}

export function* toggleTodoSaga(action : ToggleTodoRequestAction)
{
   try
   {
      const { id } = action.payload      
      const todoState : TodoState = yield select(selectTodos);
      const todo : ITodo = todoState.todos.find(t => t.id === id) as ITodo
      const toggledTodo : ITodo = { ...todo, completed : !todo.completed}
   
      yield call(patchCompleted, id, toggledTodo.completed);
      yield put(toggleTodoSuccessAction(toggledTodo));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(toggleTodoFailedAction(error.message));
   }
}

export function* removeTodoSaga(action : RemoveTodoRequestAction)
{
   try
   {
     const { id } = action.payload

     yield call(removeTodo, id);
     yield put(removeTodoSuccessAction(id));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(removeTodoFailedAction(error.message));
   }
}

function* watchTodosSagas() {
  yield takeLatest(TodoAction.GET_TODOS_REQUEST, getAllTodosSaga);
  yield takeLatest(TodoAction.ADD_TODO_REQUEST, addTodoSaga);
  yield takeLatest(TodoAction.EDIT_TODO_REQUEST, editTodoSaga);
  yield takeLatest(TodoAction.TOGGLE_TODO_REQUEST, toggleTodoSaga);
  yield takeLatest(TodoAction.REMOVE_TODO_REQUEST, removeTodoSaga);
}

export default watchTodosSagas;