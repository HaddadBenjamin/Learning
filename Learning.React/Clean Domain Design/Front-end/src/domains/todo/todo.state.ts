import { ITodo } from "./todo.model";
import { ActionStatus, IActionMetadata } from 'shared/domains/redux/redux.model';

export interface TodoState extends IActionMetadata
{
	todos : ITodo[]
}

export const initialTodoState : TodoState =
	{
		status : ActionStatus.Loading,
		error : undefined,
		todos : [],
	}
