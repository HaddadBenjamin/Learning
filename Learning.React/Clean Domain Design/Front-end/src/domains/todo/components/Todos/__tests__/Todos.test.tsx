import {useDispatch, useSelector} from "react-redux";
import {render} from "@testing-library/react";
import Todos from "../Todos";
import {todoStateMock} from "domains/todo/todo.mock";
import {getTodosRequestAction} from "domains/todo/todo.action";
import {act} from "react-dom/test-utils";
import {lazyStore} from "../../../../root/root.store";
import {todoReducerKey, todoSagaKey} from "../../../todo.configuration";
import todoReducer from "../../../todo.reducer";
import {todoSaga} from "../../../todo.saga";

jest.mock('react-redux')
const mockUseDispatch = useDispatch as jest.MockedFunction<typeof useDispatch>
const mockDispatch = jest.fn()

describe("Todos", () => {
    beforeEach(() => {
        (mockUseDispatch as jest.Mock).mockReturnValue(mockDispatch)
    })

    it("should match snapshot", async () => {
        // Given
        (useSelector as jest.Mock).mockReturnValue(todoStateMock)

        // When
        const {container} = render(<Todos/>)

        // Then
        expect(container).toMatchSnapshot()
    })
    
    it("should inject todo reducer and todo saga when component is mounted", async() =>
    {
        // Given
        const injectReducerSpy = jest.spyOn(lazyStore, "injectReducer")
        const injectSagaSpy = jest.spyOn(lazyStore, "injectSaga")
    
        // When
        await act(async () => {
            render(<Todos/>)
        })
    
        // Then
        expect(injectReducerSpy).toBeCalledWith(todoReducerKey, todoReducer)
        expect(injectSagaSpy).toBeCalledWith(todoSagaKey, todoSaga)
    })
    
    it("should dispatch getTodosRequestAction when component is mounted", async() =>
    {
        // Given
        (useSelector as jest.Mock).mockReturnValue(todoStateMock)
       
        // When
        await act(async () => {
            render(<Todos/>)
        })

        // Then
        expect(mockDispatch).toBeCalledWith(getTodosRequestAction())
    })
})