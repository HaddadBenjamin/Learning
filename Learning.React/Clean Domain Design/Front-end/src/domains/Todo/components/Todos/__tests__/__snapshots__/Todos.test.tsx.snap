// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos should match snapshot 1`] = `
<div>
  <h1
    class="title"
  >
    Todo List
  </h1>
  <div
    class="container"
  >
    <div
      class="todo"
    >
      <div>
        <input
          class="toggle"
          type="checkbox"
        />
        <input
          class="content"
          value="Faire le ménage"
        />
      </div>
      <div
        class="removeButton"
      >
        X
      </div>
    </div>
    <div
      class="todo"
    >
      <div>
        <input
          checked=""
          class="toggle"
          type="checkbox"
        />
        <input
          class="content"
          value="Faire ma compta"
        />
      </div>
      <div
        class="removeButton"
      >
        X
      </div>
    </div>
    <div
      class="todo"
    >
      <div>
        <input
          class="toggle"
          type="checkbox"
        />
        <input
          class="content"
          value="Appeler Marine"
        />
      </div>
      <div
        class="removeButton"
      >
        X
      </div>
    </div>
    <div
      class="todo"
    >
      <div>
        <input
          class="toggle"
          type="checkbox"
        />
        <input
          class="content"
          value="Acheter des croquettes pour petit chien"
        />
      </div>
      <div
        class="removeButton"
      >
        X
      </div>
    </div>
    <div
      class="todo"
    >
      <div
        class="container"
      >
        <div
          class="addButton"
        >
          +
        </div>
        <input
          class="input"
          placeholder="Ajouter une tâche"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
